src/Shared/RoslynUtils/WellKnownTypes.cs

```diff
@@ -86,6 +86,26 @@ private INamedTypeSymbol GetAndCache(int index)
         return _lazyWellKnownTypes[index]!;
     }
 
+    // Look for types within well-known assemblies only.
+    private INamedTypeSymbol? GetTypeByMetadataNameInTargetAssembly(string metadataName)
+    {
+        var types = _compilation.GetTypesByMetadataName(metadataName);
+        if (types.Length == 1)
+        {
+            return types[0];
+        }
+
+        foreach (var type in types)
+        {
+            if (type.ContainingAssembly.Identity.Name.StartsWith("System", StringComparison.Ordinal)
```

Comment:
So this will still fail if someone uses a Microsoft.* or System.* assembly name?
Is there any other heuristics we can use instead/in addition to this?

Reply:
> So this will still fail if someone uses a Microsoft.* or System.* assembly name?

Yep. If someone defines a type with the same name in an assembly that starts with `Microsoft` or `System` then this will fail.

> Is there any other heuristics we can use instead/in addition to this?

It's more tedious to apply but we could include assembly names for each type in WellKnownTypeData and use that in the lookup.

Reply:
What do analyzers and other code in dotnet/roslyn do in this situation? Surely they would have the same problem. It would be good to get a tried and tested solution.

Reply:
I see Roslyn has `GetBestTypeByMetadataName`:

https://github.com/dotnet/roslyn/blob/211e260799499391d5221ad4fa81c4e505d89871/src/Workspaces/SharedUtilitiesAndExtensions/Compiler/Core/Extensions/CompilationExtensions.cs#L36

Reply:
> What do analyzers and other code in dotnet/roslyn do in this situation? Surely they would have the same problem. It would be good to get a tried and tested solution.

I took this approach based on the guidance in the docs [here](https://learn.microsoft.com/en-us/dotnet/api/microsoft.codeanalysis.compilation.gettypebymetadataname?view=roslyn-dotnet-4.6.0#remarks). I've only observed on other instance [in one of Maui's source generators](https://grep.app/search?q=GetTypesByMetadataName&filter[repo.pattern][0]=dotnet). My bet is that most analyzers do suffer from this bug but are probably not catching it because there hasn't been a scenario where it occurred (e.g. no one has defined a `System.Task` type in their own namespace).

> I see Roslyn has GetBestTypeByMetadataName:
> https://github.com/dotnet/roslyn/blob/211e260799499391d5221ad4fa81c4e505d89871/src/Workspaces/SharedUtilitiesAndExtensions/Compiler/Core/Extensions/CompilationExtensions.cs#L36

I can give this a try. Based on the description in the code comment, it doesn't look like it'll do what we want. If anything, it seems more likely to favor types in the user's compilation as opposed to those defined in the framework.

Reply:
> I can give this a try. Based on the description in the code comment, it doesn't look like it'll do what we want. If anything, it seems more likely to favor types in the user's compilation as opposed to those defined in the framework.

Yep, this won't work for our scenario. The primary distinction between `GetBestTypeByMetadataName` and `GetTypeByMetadataName` is that the former favors types in the user's target assembly. For our scenario, where duplicate types are defined in referenced assemblies, the behavior is equivalent. It returns null and falls through to our exception.

Reply:
> What do analyzers and other code in dotnet/roslyn do in this situation? Surely they would have the same problem. It would be good to get a tried and tested solution.

[dotnet/roslyn's WellKnownTypeProvider](https://github.com/dotnet/roslyn-analyzers/blob/main/src/Utilities/Compiler/WellKnownTypeProvider.cs) implementation appears to use a strategy similar to the `GetBestTypeByMetadataName` where multiple visibile types with the same name in differing assemblies returns null.

@sharwell Thoughts on how we solve the problem of multiple types in different assemblies here? We want to favor the types defined in our assemblies, over those in others. A scenario is a third-party package which uses the same namespace + class as ours.